### PR TITLE
Add HLSL shader sample

### PR DESCRIPTION
**Pull request by SaschaWillems**
## Description

This PR adds a new (simple) sample that shows how to use HLSL shaders in a Vulkan application. It uses Glslang to convert HLSL to SPIR-V. While Glslang's HLSL support isn't as feature-rich as the one in DXC, this sample at least shows how to use basic HLSL shaders and let's people know that HLSL and Vulkan is a thing. It also comes with a tutorial/readme linking to various "HLSL in Vulkan" related resources like the Vulkan Guide entry.

## General Checklist:

Please ensure the following points are checked:

- [x] My code follows the [coding style](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Code-Style)
- [x] I have reviewed file [licenses](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#Copyright-Notice-and-License-Template)
- [x] I have updated the header Copyright to reflect the current year (CI build will fail if Copyright is out of date)
- [x] I have commented any added functions (in line with Doxygen)
- [x] I have commented any code that could be hard to understand
- [x] My changes do not add any new compiler warnings
- [x] My changes do not add any new validation layer errors or warnings
- [x] I have used existing framework/helper functions where possible
- [x] My changes build on Windows, Linux, macOS and Android. Otherwise I have [documented any exceptions](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] My changes do not add any regressions
- [x] I have tested every sample to ensure everything runs correctly
- [x] This PR describes the scope and expected impact of the changes I am making

## Sample Checklist

If your PR contains a new or modified sample, these further checks must be carried out *in addition* to the General Checklist:
- [x] I have tested the sample on at least one compliant Vulkan implementation
- [x] If the sample is vendor-specific, I have [tagged it appropriately](https://github.com/KhronosGroup/Vulkan-Samples/tree/master/CONTRIBUTING.md#General-Requirements)
- [x] I have stated on what implementation the sample has been tested so that others can test on different implementations and platforms
- [x] Any dependent assets have been merged and published in downstream modules
- [x] For new samples, I have added a paragraph with a summary to the appropriate chapter in the [samples readme](./../samples/README.md)

### CODE REVIEW

File: samples/api/hlsl_shaders/hlsl_shaders.cpp [hunk 1]
```diff
@@ -0,0 +1,513 @@
...
+	    // Binding 0 : Vertex shader uniform buffer
+	    vkb::initializers::write_descriptor_set(
+	        base_descriptor_set,
+	        VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER,
+	        0,
+	        &buffer_descriptor),
+	    // Binding 1 : Fragment shader sampled image
+	    vkb::initializers::write_descriptor_set(
+	        base_descriptor_set,
+	        VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER,
```
Comment (gary-sweet) [inline samples/api/hlsl_shaders/hlsl_shaders.cpp:355]: The descriptor set layout has this as VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE not COMBINED. This causes a validation issue:

[error] [framework/core/instance.cpp:41] 1165064310 - VUID-VkGraphicsPipelineCreateInfo-layout-00756: Validation Error: [ VUID-VkGraphicsPipelineCreateInfo-layout-00756 ] Object 0: handle = 0x55cfd1c74f90, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x45717876 | Type mismatch on descriptor slot 0.1 (expected `VK_DESCRIPTOR_TYPE_SAMPLER, VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER`) but descriptor of type VK_DESCRIPTOR_TYPE_SAMPLED_IMAGE The Vulkan spec states: layout must be consistent with all shaders specified in pStages (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-VkGraphicsPipelineCreateInfo-layout-00756)

**Review by gary-sweet — CHANGES_REQUESTED** [review]

File: samples/api/hlsl_shaders/hlsl_shaders.cpp (hunk 1 above)
Comment (SaschaWillems) [reply samples/api/hlsl_shaders/hlsl_shaders.cpp:355]: Good catch. Somehow didn't notice that one.

File: samples/api/hlsl_shaders/hlsl_shaders.cpp [hunk 2]
```diff
@@ -0,0 +1,513 @@
...
+ */
+
+/*
+ * Using HLSL shaders in Vulkan with the glslang library
+ */
+
+#include "hlsl_shaders.h"
+
+VKBP_DISABLE_WARNINGS()
+#include <SPIRV/GLSL.std.450.h>
```
Comment (gpx1000) [inline samples/api/hlsl_shaders/hlsl_shaders.cpp]: not used

File: samples/api/hlsl_shaders/hlsl_shaders.cpp [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+#include <SPIRV/GlslangToSpv.h>
+#include <StandAlone/ResourceLimits.h>
+#include <glslang/Include/ShHandle.h>
+#include <glslang/OSDependent/osinclude.h>
```
Comment (gpx1000) [inline samples/api/hlsl_shaders/hlsl_shaders.cpp]: not used

File: samples/api/hlsl_shaders/hlsl_shaders.cpp [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+	EShLanguage language{};
+	switch (stage)
+	{
+		case VK_SHADER_STAGE_VERTEX_BIT:
+			language = EShLangVertex;
+			break;
+		case VK_SHADER_STAGE_FRAGMENT_BIT:
+			language = EShLangFragment;
+			break;
+	}
```
Comment (gpx1000) [inline samples/api/hlsl_shaders/hlsl_shaders.cpp:52]: add default case

File: samples/api/hlsl_shaders/hlsl_shaders.cpp [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+
+	glslang::FinalizeProcess();
+
+	// Create shader module from generated SPIR-V
+	VkShaderModule           shader_module;
+	VkShaderModuleCreateInfo module_create_info{};
+	module_create_info.sType    = VK_STRUCTURE_TYPE_SHADER_MODULE_CREATE_INFO;
+	module_create_info.codeSize = spirv.size() * sizeof(uint32_t);
+	module_create_info.pCode    = spirv.data();
+	VK_CHECK(vkCreateShaderModule(get_device().get_handle(), &module_create_info, NULL, &shader_module));
```
Comment (gpx1000) [inline samples/api/hlsl_shaders/hlsl_shaders.cpp]: NULL -> nullptr

File: samples/api/hlsl_shaders/hlsl_shaders.cpp [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+	shader_stage.sType                           = VK_STRUCTURE_TYPE_PIPELINE_SHADER_STAGE_CREATE_INFO;
+	shader_stage.stage                           = stage;
+	shader_stage.module                          = shader_module;
+	shader_stage.pName                           = "main";
+	assert(shader_stage.module != VK_NULL_HANDLE);
+	shader_modules.push_back(shader_stage.module);
+	return shader_stage;
+}
+
+HlslShaders::HlslShaders()
```
Comment (gpx1000) [inline samples/api/hlsl_shaders/hlsl_shaders.cpp:125]: index_count, base_descriptor_set_layout, sampler_descriptor_set_layout
aren't initialized.

File: samples/api/hlsl_shaders/hlsl_shaders.cpp [hunk 7]
```diff
@@ -0,0 +1,513 @@
...
+{
+	std::vector<uint32_t> spirv;
+	std::string           info_log;
+
+	// Compile HLSL to SPIR-V
+
+	// Initialize glslang library
+	glslang::InitializeProcess();
+
+	EShMessages messages = static_cast<EShMessages>(EShMsgReadHlsl | EShMsgDefault | EShMsgVulkanRules | EShMsgSpvRules);
```
Comment (gpx1000) [inline samples/api/hlsl_shaders/hlsl_shaders.cpp]: use auto for explicit static cast.

File: samples/api/hlsl_shaders/hlsl_shaders.h [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+	VkPipeline       pipeline            = VK_NULL_HANDLE;
+	VkPipelineLayout pipeline_layout     = VK_NULL_HANDLE;
+	VkDescriptorSet  base_descriptor_set = VK_NULL_HANDLE;
+
+	VkDescriptorSetLayout base_descriptor_set_layout;
+	VkDescriptorSetLayout sampler_descriptor_set_layout;
+
+	HlslShaders();
+	~HlslShaders() override;
+	virtual void                    request_gpu_features(vkb::PhysicalDevice &gpu) override;
```
Comment (gpx1000) [inline samples/api/hlsl_shaders/hlsl_shaders.h]: virtual or override not both, prefer override

**Review by gpx1000 — DISMISSED** [review]
Nice sample Sascha!

Minor warnings came up here.  Otherwise it's good; no validation issues or other problems to my eye.

**Comment by SaschaWillems** [comment]
@gpx1000: Thanks for the feedback. I have addressed all the points you mentioned :)

File: samples/api/hlsl_shaders/hlsl_shaders.cpp (hunk 4 above)
Comment (gpx1000) [reply samples/api/hlsl_shaders/hlsl_shaders.cpp:52]: default is helpful here.

**Review by gary-sweet — APPROVED** [review]

**Review by gpx1000 — APPROVED** [review]

**Review by TomAtkinsonArm — APPROVED** [review]
LGTM!

Simple and concise, works without issue on my end

File: samples/api/hlsl_shaders/hlsl_shaders.cpp (hunk 4 above)
Comment (TomAtkinsonArm) [reply samples/api/hlsl_shaders/hlsl_shaders.cpp:52]: default could be an `assert(false && "VkShaderStageFlagBits not supported")`?

**Comment by marty-johnson59** [comment]
merging with 3 approvals

**Pull request merged**